tests/test_vllm_flash_attn.py

```diff
@@ -522,3 +639,129 @@ def test_sparse_attention_varlen(
+@pytest.mark.parametrize("causal", [False, True])
+@pytest.mark.parametrize(
+    "seqlen_q,seqlen_k",
+    [
+        (1, 1),
+        (64, 128),
+    ],
+)
+@pytest.mark.parametrize("fa_version", VERSIONS)
+def test_flash_attn_varlen_output(
+    dtype: torch.dtype,
+    num_heads: int,
+    head_size: int,
+    batch_size: int,
+    causal: bool,
+    seqlen_q: int,
+    seqlen_k: int, 
+    fa_version: int,
+):
+    device = "cuda"
+    torch.random.manual_seed(123)
+
+    q = torch.randn(batch_size, seqlen_q, num_heads, head_size,
+                    device=device, dtype=dtype)
+    k = torch.randn(batch_size, seqlen_k, num_heads, head_size,
+                    device=device, dtype=dtype)
+    v = torch.randn(batch_size, seqlen_k, num_heads, head_size,
+                    device=device, dtype=dtype)
+    query_padding_mask = generate_random_padding_mask(
+        seqlen_q, batch_size, device, mode="random", zero_lengths=False
+    )
+    key_padding_mask = generate_random_padding_mask(
+        seqlen_k, batch_size, device, mode="random", zero_lengths=True
+    )
+
+    def _gen_unused_masks(padding_mask, max_seq_len, bs, device):
+        another_mask = generate_random_padding_mask(max_seq_len, bs, device)
+        attn_mask = torch.logical_and(padding_mask, another_mask)
+        unused_mask = torch.logical_xor(
+            torch.logical_or(padding_mask, another_mask), attn_mask
+        )
+        return attn_mask, unused_mask
+
+    query_padding_mask, query_unused_mask = _gen_unused_masks(
+        query_padding_mask, seqlen_q, batch_size, q.device
+    )
+    key_padding_mask, key_unused_mask = _gen_unused_masks(
+        key_padding_mask, seqlen_k, batch_size, k.device
+    )
+
+    (
+        q_unpad,
+        k_unpad,
+        v_unpad,
+        cu_seqlens_q,
+        cu_seqlens_k,
+        _, # seqused_q
+        seqused_k,
+        max_seqlen_q,
+        max_seqlen_k,
+        q,
+        k,
+        v,
+        output_pad_fn,
+        _, # dq_pad_fn
+        _, # dk_pad_fn
+    ) = generate_qkv(q, k, v, query_padding_mask, key_padding_mask,
+                     query_unused_mask=query_unused_mask,
+                     key_unused_mask=key_unused_mask)
+    out_ref, attn_ref = attention_ref(
+        q,
+        k,
+        v,
+        query_padding_mask,
+        key_padding_mask,
+        causal=causal,
+    )
+    out_pt, attn_pt = attention_ref(
+        q,
+        k,
+        v,
+        query_padding_mask,
+        key_padding_mask,
+        causal=causal,
+        upcast=False,
+        reorder_ops=True,
+    )
+
+    print(f"Pytorch max diff: {(out_pt - out_ref).abs().max().item()}")
+    print(f"Pytorch mean diff: {(out_pt - out_ref).abs().mean().item()}")
+
+    q_zero_masking = rearrange(query_unused_mask, "b s -> b s 1 1")
+
+    # Numerical error if we just do any arithmetic on out_ref
+    fwd_atol = 2 * (out_ref + 0.3 - 0.3 - out_ref).abs().max().item()
+    rtol = 2
```

Comment:
is rtol too large? should it be 1e-2?

Reply:
This is copied from flash attention's value:

https://github.com/Dao-AILab/flash-attention/blob/main/hopper/test_flash_attn.py#L185

Reply:
Maybe worth citing in the comment?